### ISSUE TITLE
Imported default symbol names can conflict with a module’s internal symbols.

### ISSUE THREAD

**Issue by mbostock** [rollup/rollup#29]
foo.js:

```js
function foo() {
  return "I am foo!";
}

export default function() {
  return foo();
};
```

bar.js:

```js
import foo from "./foo";

foo();
```

Rollup of bar.js:

```js
function _foo() {
  return "I am foo!";
}

var _foo = function() {
  return _foo();
}

_foo();
```

This causes an infinite loop because the default import from foo into bar becomes `var _foo` which overwrites the earlier `function _foo`.

**Comment by mbostock** [comment]
Actually, it doesn’t have anything to do with the file names. Here’s another example:

foo.js:

```js
function thing() {
  return "I am thing!";
}

export default function() {
  return thing();
};
```

bar.js:

```js
import thing from "./foo";

thing();
```

This has the same problem. I can work around by renaming `thing` in foo.js to `thing_foo` or `thing` in bar.js to `thing_bar`.

**Comment by caridy** [comment]
a couple of notes here: 

* default function should be hoisted in case some other module imported it in a circular fashion.
* synthetic identifier for a default export function, or class, should probably be named accordingly, something like: `_foo_$default_` or something like that, to avoid collision with named exports in the same module.

**Comment by mbostock** [comment]
Thanks for the fast fix.

**Comment by Rich-Harris** [comment]
No problem - released as 0.8.1. @caridy Thanks for spotting that subtle point re hoisting - the new behaviour is that anonymous function expressions are converted into named function declarations (I figured that was easier than worrying about manually hoisting the definition).

**Comment by caridy** [comment]
:clap:
